content/post/2020-02-07-processx-blocked-sigchld-fix/index.Rmd

```diff
@@ -0,0 +1,442 @@
+current `SIGCHLD` handler:
+
+```
+(lldb) call (void*) malloc(sizeof(struct sigaction))                                                               (void *) $0 = 0x00007fe6c7e305f0
+(lldb) call (int) __sigaction(20, NULL, $0)
+(int) $1 = 0
+(lldb) p ((struct sigaction *)$0)->__sigaction_u                                                                   (__sigaction_u) $2 = {
+  __sa_handler = 0x000000010e8a0750 (parallel.so`parent_sig_handler)
+  __sa_sigaction = 0x000000010e8a0750 (parallel.so`parent_sig_handler)
+}
+```
+
+20 is the number of `SIGCHLD`, see `man signal`, and for the (platform
+dependent) structure of `struct sigaction` see `man sigaction`.
+In any case, `parallel.so`'s `parent_sig_handler` is indeed set up properly.
+Now we tell lldb to stop on receiving a `SIGCHLD` signal:
+
+```
+(lldb) process handle SIGCHLD --notify true --pass true --stop true                                                NAME         PASS   STOP   NOTIFY
+===========  =====  =====  ======
+SIGCHLD      true   true   true
+```
+
+and we are ready to continue the process. It will stop again at `mc_fork()`,
+because we are starting two subprocesses in `mclapply()`. Then we continue
+again:
+
+```
+(lldb) c
+Process 4196 resuming
+Process 4196 stopped
+* thread #1, queue = 'com.apple.main-thread', stop reason = breakpoint 1.1
+    frame #0: 0x000000010e89ea10 parallel.so`mc_fork
+parallel.so`mc_fork:
+->  0x10e89ea10 <+0>: pushq  %rbp
+    0x10e89ea11 <+1>: movq   %rsp, %rbp
+    0x10e89ea14 <+4>: pushq  %r15
+    0x10e89ea16 <+6>: pushq  %r14
+Target 0: (R) stopped.
+(lldb) c
+Process 4196 resuming
+Process 4196 stopped
+* thread #1, queue = 'com.apple.main-thread', stop reason = breakpoint 2.1
+    frame #0: 0x000000010e89e680 parallel.so`mc_cleanup
+parallel.so`mc_cleanup:
+->  0x10e89e680 <+0>: pushq  %rbp
+    0x10e89e681 <+1>: movq   %rsp, %rbp
+    0x10e89e684 <+4>: pushq  %r15
+    0x10e89e686 <+6>: pushq  %r14
+Target 0: (R) stopped.
+(lldb)
+```
+
+Interestingly, we got to `mc_cleanup`, even tough the `SIGCHLD` signal(s)
+should have arrived first. From another R session, we can check that the
+subprocesses of parallel are zombies already:
+
+```
+❯ sapply(ps::ps_children(ps::ps_handle(4196L)), ps::ps_status)
+[1] "zombie" "zombie"
+```
+
+We can continue still, and hope that the `SIGCHLD`s will still arrive,
+but they won't:
+
+```
+(lldb) c
+Process 4196 resuming
+Process 4196 stopped
+* thread #1, queue = 'com.apple.main-thread', stop reason = breakpoint 2.1
+    frame #0: 0x000000010e89e680 parallel.so`mc_cleanup
+parallel.so`mc_cleanup:
+->  0x10e89e680 <+0>: pushq  %rbp
+    0x10e89e681 <+1>: movq   %rsp, %rbp
+    0x10e89e684 <+4>: pushq  %r15
+    0x10e89e686 <+6>: pushq  %r14
+Target 0: (R) stopped.
+(lldb) c
+Process 4196 resuming
+Process 4196 exited with status = 0 (0x00000000)
+(lldb)
+```
+
+Clearly, the signal handler is properly set up in parallel, but the
+signals are not delivered to the process.
```

Comment:
Explicitly remind the title of the section i.e. that at this point you knew the mistake was in your code?